app/post_metadata.go

```diff
@@ -383,74 +417,154 @@ func (a *App) getImagesInMessageAttachments(post *model.Post) []string {
 	return images
 }
 
-func (a *App) getLinkMetadata(requestURL string, timestamp int64, isNewPost bool) (*opengraph.OpenGraph, *model.PostImage, error) {
+func looksLikeAPermalink(url, siteURL string) bool {
+	expression := fmt.Sprintf(`^(%s).*(/pl/)[a-z0-9]{26}$`, siteURL)
+	matched, err := regexp.MatchString(expression, strings.TrimSpace(url))
+	if err != nil {
+		mlog.Warn("error matching regex", mlog.Err(err))
+	}
+	return matched
+}
+
+func postIDFromPermalink(permalinkURL string) string {
+	return permalinkURL[len(permalinkURL)-26:]
+}
+
+func (a *App) getLinkMetadata(requestURL string, timestamp int64, isNewPost bool) (*opengraph.OpenGraph, *model.PostImage, *model.Permalink, error) {
 	requestURL = resolveMetadataURL(requestURL, a.GetSiteURL())
 
 	timestamp = model.FloorToNearestHour(timestamp)
 
 	// Check cache
-	og, image, ok := getLinkMetadataFromCache(requestURL, timestamp)
+	og, image, permalink, ok := getLinkMetadataFromCache(requestURL, timestamp)
+	if !*a.Config().ServiceSettings.EnablePermalinkPreviews {
+		permalink = nil
+	}
+
 	if ok {
-		return og, image, nil
+		return og, image, permalink, nil
 	}
 
 	// Check the database if this isn't a new post. If it is a new post and the data is cached, it should be in memory.
 	if !isNewPost {
-		og, image, ok = a.getLinkMetadataFromDatabase(requestURL, timestamp)
+		og, image, permalink, ok = a.getLinkMetadataFromDatabase(requestURL, timestamp)
+
+		if !*a.Config().ServiceSettings.EnablePermalinkPreviews {
+			permalink = nil
+		}
+
+		if permalink != nil && permalink.PreviewPost == nil {
+			postID := postIDFromPermalink(requestURL)
+
+			referencedPost, appErr := a.GetSinglePost(postID)
+			// Ignore 'not found' errors; post could have been deleted via retention policy so we don't want to permanently log a warning.
+			//
+			// TODO: Look into saving a value in the LinkMetadat.Data field to prevent perpetually re-querying for the deleted post.
+			if appErr != nil && appErr.StatusCode != http.StatusNotFound {
+				return nil, nil, nil, appErr
+			}
+
+			var referencedChannel *model.Channel
+			if referencedPost != nil {
+				referencedChannel, appErr = a.GetChannel(referencedPost.ChannelId)
+				if appErr != nil {
+					return nil, nil, nil, appErr
+				}
+			}
+
+			var referencedTeam *model.Team
+			if referencedChannel != nil {
+				referencedTeam, appErr = a.GetTeam(referencedChannel.TeamId)
+				if appErr != nil {
+					return nil, nil, nil, appErr
+				}
+			}
+
+			permalink.PreviewPost = model.PreviewPostFromPost(referencedPost, referencedTeam, referencedChannel)
```

Comment:
0/5, but perhaps we could make a store method to get a PreviewPost in a single DB query

Reply:
That's a great idea that I'd like to defer to a future improvement for the sake of getting this out an iterating on it. I've added [a Jira story](https://mattermost.atlassian.net/browse/MM-37081) that links to this comment.